precompiles/auctions/auctions.go

```diff
@@ -0,0 +1,136 @@
+
+import (
+	"embed"
+	"fmt"
+
+	erc20keeper "github.com/evmos/evmos/v19/x/erc20/keeper"
+
+	auctionskeeper "github.com/evmos/evmos/v19/x/auctions/keeper"
+
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	authzkeeper "github.com/cosmos/cosmos-sdk/x/authz/keeper"
+	"github.com/ethereum/go-ethereum/common"
+	contractutils "github.com/evmos/evmos/v19/contracts/utils"
+	cmn "github.com/evmos/evmos/v19/precompiles/common"
+	"github.com/evmos/evmos/v19/x/evm/core/vm"
+)
+
+var _ vm.PrecompiledContract = &Precompile{}
+
+const PrecompileAddress string = "0x0000000000000000000000000000000000000805"
+
+// Embed abi json file to the executable binary. Needed when importing as dependency.
+//
+//go:embed abi.json
+var f embed.FS
+
+// Precompile defines the precompiled contract for auctions.
+type Precompile struct {
+	cmn.Precompile
+	auctionsKeeper auctionskeeper.Keeper
+	erc20Keeper    erc20keeper.Keeper
+}
+
+// NewPrecompile creates a new auctions Precompile instance as a
+// PrecompiledContract interface.
+func NewPrecompile(
+	auctionsKeeper auctionskeeper.Keeper,
+	erc20Keeper erc20keeper.Keeper,
+	authzKeeper authzkeeper.Keeper,
+) (*Precompile, error) {
+	newAbi, err := contractutils.LoadABI(f, "abi.json")
+	if err != nil {
+		return nil, fmt.Errorf("error loading the auction ABI %s", err)
+	}
+
+	p := &Precompile{
+		Precompile: cmn.Precompile{
+			ABI:                  newAbi,
+			AuthzKeeper:          authzKeeper,
+			KvGasConfig:          storetypes.KVGasConfig(),
+			TransientKVGasConfig: storetypes.TransientGasConfig(),
+			ApprovalExpiration:   cmn.DefaultExpirationDuration, // should be configurable in the future.
+		},
+		erc20Keeper:    erc20Keeper,
+		auctionsKeeper: auctionsKeeper,
+	}
+
+	// SetAddress defines the address of the auction compile contract.
+	p.SetAddress(common.HexToAddress(PrecompileAddress))
+
+	return p, nil
+}
+
+// RequiredGas calculates the precompiled contract's base gas rate.
+func (p Precompile) RequiredGas(input []byte) uint64 {
+	// NOTE: This check avoid panicking when trying to decode the method ID
+	if len(input) < 4 {
+		return 0
+	}
+
+	methodID := input[:4]
+
+	method, err := p.MethodById(methodID)
+	if err != nil {
+		// This should never happen since this method is going to fail during Run
+		return 0
+	}
+
+	return p.Precompile.RequiredGas(input, p.IsTransaction(method.Name))
+}
+
+// Run executes the precompiled contract auction methods defined in the ABI.
+func (p Precompile) Run(evm *vm.EVM, contract *vm.Contract, readOnly bool) (bz []byte, err error) {
+	ctx, stateDB, snapshot, method, initialGas, args, err := p.RunSetup(evm, contract, readOnly, p.IsTransaction)
+	if err != nil {
+		return nil, err
+	}
+
+	// This handles any out of gas errors that may occur during the execution of a precompile tx or query.
+	// It avoids panics and returns the out of gas error so the EVM can continue gracefully.
+	defer cmn.HandleGasError(ctx, contract, initialGas, &err)()
+
+	switch method.Name {
```

Comment:
should there be a query about my previous bid in the auction?

Reply:
That would require storing more data in store, the current approach is to filter through events in which the sender and the round are indexed so it would be easy to query this with the filter.

Reply:
not necessarily, just for the current bid and then you delete them when the epoch ends. This is the same approach that the governance uses for votes

Reply:
Still that's extra operations at the end of the epoch. Using the event filter you can easily filter by sender address and round when the user connects their wallet and get all the bids they made for that round.